{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "RzIc-rg5CHP9"
      },
      "source": [
        "#Protein Metrics\n",
        "\n",
        "This notebook calculates various sequence- and structure-based quality scores for proteins, such as those produced by generative models. Many different kinds of metrics can be calculated. For all of the metrics, proteins with higher scores (closer to zero for negative numbers) are predicted to be more likely to fold and have function than proteins with lower scores.\n",
        "\n",
        "**inputs**: Please provide protein sequences in fasta format, AlphaFold2-predicted structures, and/or reference protein sequences in fasta format as appropriate (see sections below).\n",
        "\n",
        "**outputs**: A comma-separated values (csv) text file with the calculated metrics.\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "LSWLh0lnAu36"
      },
      "source": [
        "## Setup"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "tkjy0gbXI_Ji"
      },
      "outputs": [],
      "source": [
        "%%capture cap --no-stderr\n",
        "#@title 1. Install dependencies and set up directories\n",
        "\n",
        "# set up directories\n",
        "!rm -r sample_data\n",
        "!mkdir -p pdbs target_seqs reference_seqs\n",
        "\n",
        "!mkdir -p /opt/bin \n",
        "\n",
        "#install esm\n",
        "!pip -q install  git+https://github.com/facebookresearch/esm.git@e5e7b06b9a093706607c229ab1c5c9821806814d \n",
        "\n",
        "#install hmmer, mafft, emboss, biotite, and sequence-models\n",
        "!apt-get -qq update > /dev/null\n",
        "!apt-get -qq install  hmmer mafft emboss > /dev/null\n",
        "!pip -q install  biotite pytest-datadir sequence-models > /dev/null\n",
        "\n",
        "#install fasta3\n",
        "!wget https://github.com/wrpearson/fasta36/releases/download/v36.3.8i_14-Nov-2020/fasta-36.3.8i-linux64.tar.gz > /dev/null\n",
        "!tar xf fasta-36.3.8i-linux64.tar.gz > /dev/null\n",
        "!mv fasta-36.3.8i/bin/* /opt/bin \n",
        "!rm -r fasta*\n",
        "\n",
        "### FROM ESM-IF example notebook https://github.com/facebookresearch/esm/blob/e5e7b06b9a093706607c229ab1c5c9821806814d/examples/inverse_folding/notebook.ipynb\n",
        "import torch\n",
        "def format_pytorch_version(version):\n",
        "  return version.split('+')[0]\n",
        "\n",
        "TORCH_version = torch.__version__\n",
        "TORCH = format_pytorch_version(TORCH_version)\n",
        "\n",
        "def format_cuda_version(version):\n",
        "  return 'cu' + version.replace('.', '')\n",
        "\n",
        "CUDA_version = torch.version.cuda\n",
        "CUDA = format_cuda_version(CUDA_version)\n",
        "\n",
        "!pip install -q torch-scatter -f https://data.pyg.org/whl/torch-{TORCH}+{CUDA}.html > /dev/null\n",
        "!pip install -q torch-sparse -f https://data.pyg.org/whl/torch-{TORCH}+{CUDA}.html > /dev/null\n",
        "!pip install -q torch-cluster -f https://data.pyg.org/whl/torch-{TORCH}+{CUDA}.html > /dev/null\n",
        "!pip install -q torch-spline-conv -f https://data.pyg.org/whl/torch-{TORCH}+{CUDA}.html > /dev/null\n",
        "!pip install -q torch-geometric > /dev/null\n",
        "!pip install -q pgen > /dev/null\n",
        "### END OF CODE FROM ESM-IF\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "nbLdSpLNtelA"
      },
      "outputs": [],
      "source": [
        "#@title 2. Reset calculated metrics (creates a new datastructure to store results, clearing any existing results)\n",
        "\n",
        "results = dict()\n",
        "device = 'cuda:0'\n",
        "\n",
        "from google.colab import files\n",
        "import os\n",
        "import re\n",
        "import torch\n",
        "\n",
        "#from sequence_models.utils import parse_fasta\n",
        "from typing import List\n",
        "from pgen.utils import parse_fasta\n",
        "import numpy as np\n",
        "import pandas as pd\n",
        "from scipy.spatial.distance import pdist\n",
        "from pathlib import Path\n",
        "from glob import glob\n",
        "from random import randrange\n",
        "import esm\n",
        "from biotite.structure.io import pdb\n",
        "from biotite.structure import get_residue_count\n",
        "from biotite.structure.residues import get_residues\n",
        "from biotite.sequence import ProteinSequence\n",
        "import tempfile\n",
        "from Bio.Align import substitution_matrices\n",
        "import subprocess\n",
        "from dataclasses import dataclass\n",
        "\n",
        "from sequence_models.pretrained import load_model_and_alphabet\n",
        "from sequence_models.pdb_utils import parse_PDB, process_coords\n",
        "\n",
        "def residues_in_pdb(pdb_path):\n",
        "    with open(pdb_path) as f:\n",
        "        pdb_file = pdb.PDBFile.read(pdb_path)\n",
        "        atoms  = pdb_file.get_structure()\n",
        "    return get_residue_count(atoms)\n",
        "\n",
        "gap_open=10\n",
        "gap_extend=2\n",
        "\n",
        "def add_metric(metrics_dict, protein_name, metric_name, value):\n",
        "  if protein_name not in metrics_dict:\n",
        "    metrics_dict[protein_name] = dict()\n",
        "  metrics_dict[protein_name][metric_name] = value\n",
        "\n",
        "def get_pdb_sequence(pdb_path):\n",
        "    with open(pdb_path) as f:\n",
        "        pdb_file = pdb.PDBFile.read(pdb_path)\n",
        "        atoms  = pdb_file.get_structure()\n",
        "        residues = get_residues(atoms)[1]\n",
        "    return ''.join([ProteinSequence.convert_letter_3to1(r) for r in residues])\n",
        "\n",
        "with open(\"/tmp/PFASUM15.mat\", \"w\") as f:\n",
        "  f.write(\"\"\"    A   R   N   D   C   Q   E   G   H   I   L   K   M   F   P   S   T   W   Y   V   B   Z   J   X   *\n",
        "A   3  -1  -1  -1   0  -1   0   1  -1   0   0  -1  -1  -2   0   1   0  -2  -2   1  -7  -6  -6 -26  -6\n",
        "R  -1   6   1   0  -3   2   1  -1   0  -4  -3   4  -2  -4   0   0   1  -3  -3  -3  -5  -4 -10 -26  -6\n",
        "N  -1   1   6   3  -3   2   1   1   3  -5  -4   1  -1  -4   0   2   0  -5  -2  -4  -4  -4 -10 -26  -6\n",
        "D  -1   0   3   8  -3   1   3   0   1  -5  -5   1  -5  -5   0   1   0  -5  -4  -4  -3  -3 -11 -27  -6\n",
        "C   0  -3  -3  -3  16  -2  -1   0  -1   0   0  -4   2   1  -2   0   0   0  -2   1  -9  -7  -6 -26  -6\n",
        "Q  -1   2   2   1  -2   5   3   0   1  -4  -4   2  -2  -3  -1   1   1  -1  -3  -3  -5  -5 -10 -26  -6\n",
        "E   0   1   1   3  -1   3   6   0   1  -5  -4   3  -3  -6   0   0   0  -3  -3  -4  -4  -5 -10 -26  -6\n",
        "G   1  -1   1   0   0   0   0   9   0  -4  -4   0  -2  -3   0   1  -1  -2  -4  -3  -5  -6 -10 -27  -6\n",
        "H  -1   0   3   1  -1   1   1   0  10  -4  -3   1  -2  -1  -1   0   0   0   2  -2  -4  -5  -9 -26  -6\n",
        "I   0  -4  -5  -5   0  -4  -5  -4  -4   5   4  -4   3   3  -4  -3  -1  -1   0   5 -11 -10  -3 -26  -6\n",
        "L   0  -3  -4  -5   0  -4  -4  -4  -3   4   5  -4   2   3  -3  -3  -2   1   1   3 -10 -10  -5 -27  -6\n",
        "K  -1   4   1   1  -4   2   3   0   1  -4  -4   5  -2  -4   0   1   0  -4  -2  -4  -5  -3 -10 -26  -6\n",
        "M  -1  -2  -1  -5   2  -2  -3  -2  -2   3   2  -2   8   2  -2  -2   0   2   1   2  -9  -9  -4 -26  -6\n",
        "F  -2  -4  -4  -5   1  -3  -6  -3  -1   3   3  -4   2   8  -3  -4  -2   5   5   1 -11 -11  -3 -26  -6\n",
        "P   0   0   0   0  -2  -1   0   0  -1  -4  -3   0  -2  -3  12   0   0  -3  -2  -3  -6  -6 -10 -27  -6\n",
        "S   1   0   2   1   0   1   0   1   0  -3  -3   1  -2  -4   0   4   2  -3  -2  -2  -5  -5  -9 -26  -6\n",
        "T   0   1   0   0   0   1   0  -1   0  -1  -2   0   0  -2   0   2   3  -3  -1   0  -6  -6  -7 -26  -6\n",
        "W  -2  -3  -5  -5   0  -1  -3  -2   0  -1   1  -4   2   5  -3  -3  -3  20   6  -2 -11  -8  -6 -27  -6\n",
        "Y  -2  -3  -2  -4  -2  -3  -3  -4   2   0   1  -2   1   5  -2  -2  -1   6  10   0  -9  -9  -6 -26  -6\n",
        "V   1  -3  -4  -4   1  -3  -4  -3  -2   5   3  -4   2   1  -3  -2   0  -2   0   5 -10 -10  -2 -26  -6\n",
        "B  -7  -5  -4  -3  -9  -5  -4  -5  -4 -11 -10  -5  -9 -11  -6  -5  -6 -11  -9 -10  -5 -10 -17 -32  -6\n",
        "Z  -6  -4  -4  -3  -7  -5  -5  -6  -5 -10 -10  -3  -9 -11  -6  -5  -6  -8  -9 -10 -10  -6 -16 -32  -6\n",
        "J  -6 -10 -10 -11  -6 -10 -10 -10  -9  -3  -5 -10  -4  -3 -10  -9  -7  -6  -6  -2 -17 -16  -6 -33  -6\n",
        "X -26 -26 -26 -27 -26 -26 -26 -27 -26 -26 -27 -26 -26 -26 -27 -26 -26 -27 -26 -26 -32 -32 -33 -40  -6\n",
        "*  -6  -6  -6  -6  -6  -6  -6  -6  -6  -6  -6  -6  -6  -6  -6  -6  -6  -6  -6  -6  -6  -6  -6  -6   1\"\"\")\n",
        "\n",
        "with open(\"/tmp/BLOSUM62.mat\", \"w\") as f:\n",
        "  f.write(\"\"\"   A  R  N  D  C  Q  E  G  H  I  L  K  M  F  P  S  T  W  Y  V  B  Z  X  *\n",
        "A  4 -1 -2 -2  0 -1 -1  0 -2 -1 -1 -1 -1 -2 -1  1  0 -3 -2  0 -2 -1  0 -4 \n",
        "R -1  5  0 -2 -3  1  0 -2  0 -3 -2  2 -1 -3 -2 -1 -1 -3 -2 -3 -1  0 -1 -4 \n",
        "N -2  0  6  1 -3  0  0  0  1 -3 -3  0 -2 -3 -2  1  0 -4 -2 -3  3  0 -1 -4 \n",
        "D -2 -2  1  6 -3  0  2 -1 -1 -3 -4 -1 -3 -3 -1  0 -1 -4 -3 -3  4  1 -1 -4 \n",
        "C  0 -3 -3 -3  9 -3 -4 -3 -3 -1 -1 -3 -1 -2 -3 -1 -1 -2 -2 -1 -3 -3 -2 -4 \n",
        "Q -1  1  0  0 -3  5  2 -2  0 -3 -2  1  0 -3 -1  0 -1 -2 -1 -2  0  3 -1 -4 \n",
        "E -1  0  0  2 -4  2  5 -2  0 -3 -3  1 -2 -3 -1  0 -1 -3 -2 -2  1  4 -1 -4 \n",
        "G  0 -2  0 -1 -3 -2 -2  6 -2 -4 -4 -2 -3 -3 -2  0 -2 -2 -3 -3 -1 -2 -1 -4 \n",
        "H -2  0  1 -1 -3  0  0 -2  8 -3 -3 -1 -2 -1 -2 -1 -2 -2  2 -3  0  0 -1 -4 \n",
        "I -1 -3 -3 -3 -1 -3 -3 -4 -3  4  2 -3  1  0 -3 -2 -1 -3 -1  3 -3 -3 -1 -4 \n",
        "L -1 -2 -3 -4 -1 -2 -3 -4 -3  2  4 -2  2  0 -3 -2 -1 -2 -1  1 -4 -3 -1 -4 \n",
        "K -1  2  0 -1 -3  1  1 -2 -1 -3 -2  5 -1 -3 -1  0 -1 -3 -2 -2  0  1 -1 -4 \n",
        "M -1 -1 -2 -3 -1  0 -2 -3 -2  1  2 -1  5  0 -2 -1 -1 -1 -1  1 -3 -1 -1 -4 \n",
        "F -2 -3 -3 -3 -2 -3 -3 -3 -1  0  0 -3  0  6 -4 -2 -2  1  3 -1 -3 -3 -1 -4 \n",
        "P -1 -2 -2 -1 -3 -1 -1 -2 -2 -3 -3 -1 -2 -4  7 -1 -1 -4 -3 -2 -2 -1 -2 -4 \n",
        "S  1 -1  1  0 -1  0  0  0 -1 -2 -2  0 -1 -2 -1  4  1 -3 -2 -2  0  0  0 -4 \n",
        "T  0 -1  0 -1 -1 -1 -1 -2 -2 -1 -1 -1 -1 -2 -1  1  5 -2 -2  0 -1 -1  0 -4 \n",
        "W -3 -3 -4 -4 -2 -2 -3 -2 -2 -3 -2 -3 -1  1 -4 -3 -2 11  2 -3 -4 -3 -2 -4 \n",
        "Y -2 -2 -2 -3 -2 -1 -2 -3  2 -1 -1 -2 -1  3 -3 -2 -2  2  7 -1 -3 -2 -1 -4 \n",
        "V  0 -3 -3 -3 -1 -2 -2 -3 -3  3  1 -2  1 -1 -2 -2  0 -3 -1  4 -3 -2 -1 -4 \n",
        "B -2 -1  3  4 -3  0  1 -1  0 -3 -4  0 -3 -3 -2  0 -1 -4 -3 -3  4  1 -1 -4 \n",
        "Z -1  0  0  1 -3  3  4 -2  0 -3 -3  1 -1 -3 -1  0 -1 -3 -2 -2  1  4 -1 -4 \n",
        "X  0 -1 -1 -1 -2 -1 -1 -1 -1 -1 -1 -1 -1 -1 -2  0  0 -2 -1 -1 -1 -1 -1 -4 \n",
        "* -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4  1\n",
        "\"\"\")\n",
        "\n",
        "with open(\"/tmp/extract_mif.py\", \"w\") as f:\n",
        "  f.write(\"\"\"import argparse\n",
        "\n",
        "import torch.cuda\n",
        "from tqdm import tqdm\n",
        "import pandas as pd\n",
        "import numpy as np\n",
        "\n",
        "from sequence_models.utils import parse_fasta\n",
        "from sequence_models.pretrained import load_model_and_alphabet\n",
        "from sequence_models.pdb_utils import parse_PDB, process_coords\n",
        "\n",
        "\n",
        "parser = argparse.ArgumentParser()\n",
        "parser.add_argument('model')\n",
        "parser.add_argument('tsv_fpath')\n",
        "parser.add_argument('out_dir')\n",
        "parser.add_argument('result')\n",
        "parser.add_argument('--include', nargs='*', default=['mean'])\n",
        "parser.add_argument('--device', default=None)\n",
        "args = parser.parse_args()\n",
        "\n",
        "# Check inputs\n",
        "if args.model not in ['mif', 'mifst']:\n",
        "    raise ValueError(\"Valid models ars 'mif' and 'mifst'.\")\n",
        "if args.result == 'logits':\n",
        "    for inc in args.include:\n",
        "        if inc not in ['per_tok', 'logp']:\n",
        "            raise ValueError(\"logits can be included as 'per_tok' or as 'logp'.\")\n",
        "elif args.result == 'repr':\n",
        "    for inc in args.include:\n",
        "        if inc not in ['per_tok', 'mean']:\n",
        "            raise ValueError(\"repr can be included as 'per_tok' or as 'mean'.\")\n",
        "else:\n",
        "    raise ValueError(\"Valid results ars 'repr' and 'logits'.\")\n",
        "\n",
        "# load model\n",
        "print('Loading model...')\n",
        "model, collater = load_model_and_alphabet(args.model)\n",
        "# detect device and move model to it\n",
        "if args.device is None:\n",
        "    if torch.cuda.is_available():\n",
        "        device = torch.device('cuda:0')\n",
        "    else:\n",
        "        device = torch.device('cpu')\n",
        "else:\n",
        "    device = torch.device(args.device)\n",
        "model = model.to(device)\n",
        "# load data\n",
        "print('Loading data...')\n",
        "df = pd.read_table(args.tsv_fpath).reset_index()\n",
        "if 'logp' in args.include:\n",
        "    logps = np.empty(len(df))\n",
        "with tqdm(total=len(df)) as pbar:\n",
        "    for i, row in df.iterrows():\n",
        "        seq = row['sequence']\n",
        "        pdb = row['pdb']\n",
        "        name = row['name']\n",
        "        coords, wt, _ = parse_PDB(pdb)\n",
        "        coords = {\n",
        "            'N': coords[:, 0],\n",
        "            'CA': coords[:, 1],\n",
        "            'C': coords[:, 2]\n",
        "        }\n",
        "        dist, omega, theta, phi = process_coords(coords)\n",
        "        batch = [[seq, torch.tensor(dist, dtype=torch.float),\n",
        "                  torch.tensor(omega, dtype=torch.float),\n",
        "                  torch.tensor(theta, dtype=torch.float), torch.tensor(phi, dtype=torch.float)]]\n",
        "        src, nodes, edges, connections, edge_mask = collater(batch)\n",
        "        src = src.to(device)\n",
        "        nodes = nodes.to(device)\n",
        "        edges = edges.to(device)\n",
        "        connections = connections.to(device)\n",
        "        edge_mask = edge_mask.to(device)\n",
        "        rep = model(src, nodes, edges, connections, edge_mask, result=args.result)[0]\n",
        "        if args.result == 'repr':\n",
        "            if 'mean' in args.include:\n",
        "                torch.save(rep.mean(dim=0).detach().cpu(),\n",
        "                           args.out_dir + '_'.join([name, args.model, 'mean']) + '.pt')\n",
        "            if 'per_tok' in args.include:\n",
        "                torch.save(rep.detach().cpu(),\n",
        "                           args.out_dir + '_'.join([name, args.model, 'per_tok']) + '.pt')\n",
        "        else:\n",
        "            if 'logp' in args.include:\n",
        "                rep = rep.log_softmax(dim=-1)\n",
        "                logps[i] = rep[torch.arange(len(src[0])), src].mean().detach().cpu().numpy()\n",
        "            if 'per_tok' in args.include:\n",
        "                torch.save(rep.detach().cpu(), args.out_dir + '_'.join([name, args.model, 'logits']) + '.pt')\n",
        "        pbar.update(1)\n",
        "    if 'logp' in args.include:\n",
        "        df['logp'] = logps\n",
        "        out_fpath = args.out_dir + args.model + '_logp.tsv'\n",
        "        print('Writing results to ' + out_fpath)\n",
        "        df = df.to_csv(out_fpath, index=False, sep=\"\\t\")\n",
        "  \"\"\")\n",
        "\n",
        "with open(\"/tmp/extract.py\", \"w\") as f:\n",
        "  f.write(\"\"\"import argparse\n",
        "\n",
        "import torch.cuda\n",
        "from tqdm import tqdm\n",
        "import numpy as np\n",
        "import pandas as pd\n",
        "\n",
        "from sequence_models.utils import parse_fasta\n",
        "from sequence_models.pretrained import load_model_and_alphabet\n",
        "\n",
        "parser = argparse.ArgumentParser()\n",
        "parser.add_argument('model')\n",
        "parser.add_argument('in_fpath')\n",
        "parser.add_argument('out_dir')\n",
        "parser.add_argument('--repr_layers', nargs='*', default=[-1])\n",
        "parser.add_argument('--include', nargs='*', default=['mean'])\n",
        "parser.add_argument('--device', default=None)\n",
        "parser.add_argument('--batchsize', default=1, type=int)\n",
        "args = parser.parse_args()\n",
        "\n",
        "# load model\n",
        "print('Loading model...')\n",
        "model, collater = load_model_and_alphabet(args.model)\n",
        "# detect device and move model to it\n",
        "if args.device is None:\n",
        "    if torch.cuda.is_available():\n",
        "        device = torch.device('cuda:0')\n",
        "    else:\n",
        "        device = torch.device('cpu')\n",
        "else:\n",
        "    device = torch.device(args.device)\n",
        "model = model.to(device)\n",
        "# load data\n",
        "print('Loading data...')\n",
        "seqs, names = parse_fasta(args.in_fpath, return_names=True)\n",
        "ells = [len(s) for s in seqs]\n",
        "seqs = [[s] for s in seqs]\n",
        "n_total = len(seqs)\n",
        "repr_layers = []\n",
        "for r in args.repr_layers:\n",
        "    if r == 'logits':\n",
        "        logits = True\n",
        "    else:\n",
        "        repr_layers.append(int(r))\n",
        "if 'logp' in args.include:\n",
        "    logps = np.empty(len(seqs))\n",
        "with tqdm(total=n_total) as pbar:\n",
        "    for i in range(0, n_total, args.batchsize):\n",
        "        start = i\n",
        "        end = start + args.batchsize\n",
        "        bs = seqs[start:end]\n",
        "        bn = names[start:end]\n",
        "        bl = ells[start:end]\n",
        "        # tokenize\n",
        "        x = collater(bs)[0].to(device)\n",
        "        # pass through the model\n",
        "        results = model(x, repr_layers=repr_layers, logits=logits)\n",
        "        if 'representations' in results:\n",
        "            for layer, rep in results['representations'].items():\n",
        "                for r, ell, name in zip(rep, bl, bn):\n",
        "                    r = r[:ell]\n",
        "                    if 'mean' in args.include:\n",
        "                        torch.save(r.mean(dim=0).detach().cpu(),\n",
        "                                   args.out_dir + '_'.join([name, args.model, str(layer), 'mean']) + '.pt')\n",
        "                    if 'per_tok' in args.include:\n",
        "                        torch.save(r.detach().cpu(),\n",
        "                                   args.out_dir + '_'.join([name, args.model, str(layer), 'per_tok']) + '.pt')\n",
        "        if logits:\n",
        "            rep = results['logits']\n",
        "            for r, ell, name, src in zip(rep, bl, bn, x):\n",
        "                if 'per_tok' in args.include:\n",
        "                    r = r[:ell]\n",
        "                    torch.save(r.detach().cpu(), args.out_dir + '_'.join([name, args.model, 'logits']) + '.pt')\n",
        "                if 'logp' in args.include:\n",
        "                    r = r.log_softmax(dim=-1)[:ell]\n",
        "                    # print(src.shape, r.shape)\n",
        "                    logps[i] = r[torch.arange(len(src)), src].mean().detach().cpu().numpy()\n",
        "        pbar.update(len(bs))\n",
        "    if 'logp' in args.include:\n",
        "        df = pd.DataFrame()\n",
        "        df['name'] = names\n",
        "        df['sequence'] = [s[0] for s in seqs]\n",
        "        df['logp'] = logps\n",
        "        out_fpath = args.out_dir + args.model + '_logp.tsv'\n",
        "        print('Writing results to ' + out_fpath)\n",
        "        df = df.to_csv(out_fpath, index=False, sep=\"\\t\")\n",
        "\"\"\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Mr-DKOrIHR3k"
      },
      "source": [
        "### 3. Upload files\n",
        "\n",
        "Different files are required depending on which kinds of metrics you want to calculate.\n",
        "\n",
        "Please upload protein PDB files or protein fasta files to the appropriate directories in `/content`\n",
        "\n",
        "- Structures as PDB files (used for structure metrics). \n",
        "   - In this notebook we assume every strucure has a single chain, chain A.\n",
        "   - pLDDT score assumes that the scores are stored in the b-factor field of the PDB file, which is the case for PDB files generated by AlphaFold2.\n",
        "- Target sequences (used for single-sequence and alignment-based metrics)\n",
        "  - If the directory contains more than one fasta file, metrics will be calculated for all sequences from all files in the directory.\n",
        "- Reference sequences (used for alignment-based metrics)\n",
        "  - If the directory contains more than one fasta file, all of them will be used as reference sequences.\n",
        "\n",
        "![image.png](data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAVgAAAEICAYAAAAa4uy3AAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAAJcEhZcwAADsMAAA7DAcdvqGQAACGkSURBVHhe7d0PcBvlnTfwLyFUTCHmkuA0vaIBHDXB6NIxpsV14qFmaByuU9tc4jO8Y8wNskNIYw7n8hKCKTV+ezgmnMF3TeoBHFNwPfPiOn2xc+XFTdP4aGKTgzgegpImVTwtyh3UapqpO+mLpknz7m+166wVSbZkPdYffz8zO9JqV9IjW/vVT8+u9rniogZERBSzX/3qV/jiF79ozF0yx7gkIqI4YwVLREr87ne/w8GDB4256C1cuBC33HILrr/+euOW5BWugmXAEpESW7Zs0UN2OiRct2/fbsxNTp5P1pfnniyYZd2enh6UlpZOO8QZsEQ0YyS8JORWrlypB1gksl4kEphTDUAz1GX9SCEr67S3t+OXv/ylXiVP1obJsA+WiGacBFykKd7MUDUrWbkMZg1XWXe64RoJA5aI0oYZmOFCNjhco+l+iAUDlogSSroQQk3y1T0W4UJ2psNVpHQf7NjJQbhHjZkJbLjhK7mwX+lFf3sXPEvKUX23HfB7MfTeafgXOZG/NMNYl4jiTcJMQs4My1jIDiiZJAglEKNlDVe5v0yqwjWuO7nkRZ84cQIul0tvvFyfCjnsoqCgwJibvqGmLJS9bMxMUIimgXaUn25EVnmbNl+N7pE65H7SBdeKreh/uBsjW3MDqxJR3CVDwApryAoV4SriupNLjm2TTwIzXM0/xGTTVIM4OpVoPzyM4QlTC0rl/5GzDp3f3YiGznVgnBLNTsHhbIbtTIipgpVwFbH2kcRLoII1qlPjtmD+s2PwfyYDGddoM+Eq2HNeuI+cxhhsyHTmwjHfuN3C/xs3hrxjwNWZcCx3INNmLCCiyyRDBSttsPa5mrfJdbOPNl7iWsFKsCY6XKdmCM235yDne0PGfDA/PJ0bkLP8ayh+sAIVD5ah6PZslLW6tSWG81701uQg+65ibbm2TnkR8m4vRvP742sQUZIJDlcJ6VA7vlSL+SiCmWjc1HgxdGAQg9bppFZpTsHYW0+g+Ol+OB7txH8cH8HI8UPYU58P9/NleOIt4zGGO/Trhd/9KY6f1NY5vAd1hTZ4P3BrFS8RJZtQ4SqsletMhWxMAWt+GkjjDhw4oM9PZZJyP/760KhXn5bpxx5jWSRe9Ozqhf+2Ory4KR92+cpvy4TzH+pQt9yP3r7BQIBeafQF+P2Bqna+E9U7utHiygWPQyBKLuHC1TTTIRtTwMbaIDmKIP5C7OR6dAq7tM6fhveIdnmkHl/LykLW+FSE+qPa7fs+hB7Ty8vRcr8Tg/9cjJylOSh6sBbNe7Tq9bwsJKJIJCtinWIhRVy4cDUFh6wEsioxHwcrDZMGJtLkO7mG0JhVhjZzp5Z1J9f/9KNxaQXa/qEdw/8Y6t42ZMy/tCfLf9qNwXcHsb/vFXTt8wF3t+CdV0qQaSwnokvM4JKwm65oAtB8XgnQyZjrlpSUTHufUlqe7GVaAbvVji5XHrZ6tPvv1e4/nqV+eA64cd1Xc5E5V5s9Pwb3f7iR8bV82GVeM/RiHsq+5wwca7s4cBsRTSQBJgF75swZ45boxXoEwkwLF7CQgI2Wz+e7ePz4cWMucQ5vu/nizTc/e/GwMX+5wxefvVlbZ5uxxsdvXHzIMv/pe89eXKXN3/L3/3Kx571fXfzIM3Dx9Se/efGWm2+5+NCPPtLX+eiHa7XnuPniqsdfvzjgGb340QdvXPzO17XHcH7n4sCn+ipENMudPHnSuDZRTBWs/IJLSB+H/Ohgqjuv5NRlVVVVxtz0Ta+CDdzDN7gTtTXNGDyrzwKfzUTJk514rsIBvag978Ngay02vGjs9BLz81H3g3ZUL7/UhUBEs1fcfyor5b+EpXwFmOovtJYtW5acx8+e92Psj4HjWm3zMmAzugIm8I9h7E9yZWLfLBFRWvbBEhElAwnYUGL+oQEREV1is9nGp6uvvlqfGLBERHHwmc98Zjxg5bpMDFgiojgwA9asXlnBEhHFibVyNcOWAUtEFAdz587FVVddNT7JPAOWiCgO5syZM2G68sorGbBERPFwxRVXXDYxYImI4kzCVTBgiYjiQO9znTNnPFwFA5aISBEGLBGRIgxYIiJFGLBERIowYImIFGHAEhEpwoAlIlKEAUtEpAgDlohIEQYsEZEiDFgiIkVm7aCHMhpub2+vfhlJaWmpPhERhRNuVNlZW8G2t7dPGq5ChiiXiWYHeU9s375dH5aeaLpmbQXrcrn0SwnacMx1RKIqWdngpY3hNvhkrbBTtd1btmzR23zLLbfo///rr7/eWBJw4MAB/fLMmTNYtmyZvjx4HZp9wlWwVz6jMa7PKmZVGmkjt1auJ06c0C9lw5tJElIfffSRMXe5RLVrMqna7lWrVuHgwYN6271erx6in/3sZ/X3glS2R44c0Sdpv6y3d+9efdnChQv19WSi2ef3v/+9/h4Ixgp2ihWsaaYrr1BtiIVUWfJYMxVoqdpuU6hKdteuXXqoRmK2lVXt7MI+2DiRasVa2aYKCYtIHyaJJGEkoRQqmBLVbrMd1q6OqqoqPUClkpVJQjg4+GVdCWFZn4gVbJQVrGmmNvqptHMq4vU4UxXpbxdMwsoMNAkmCa5gM9VuIe0J3gFqrWSDhfrQXblypf5NJ9T6lH5YwcZANurgieJLQsgaThJIEk6JIm0JDlcRvNPODFWZDxWkUsVO1p1A6Y8BSwkjoSThJDuSrKE2k33cVhKWwZWoMAPUGrJym1yaIVpSUqJfWpkBTLMXAzYEORQn1IaWypKxAje7EeTrt7XakzCLposhXsL9fSRMpdtAWENW+mTNAA3uizWl2/uIosOADSLhKhuQbBipvHFI22Wvt0zmsZvmzplkIIFkhpIEqlSxZjuFdXmimSFqdgMEh2wkcrwszV5xC9h0+CpkhqspFUNWNn6p/qTdcqymTPKazMOOkkVwhSo/W5bJJGEW6mu3StImM0StzO4La9+wNWRFuL9tIipxSh7TClh5U0mFJG8i2YBlkgopFcM2OFyFbGypFrLyGqTd5v9CLuX/I/+TZPm/mH2aJvnbm+0z2y2TNXBngrQp1A62gYEB/VLaLcvNtltDVqrt4PsGv06afWIOWHlzycYgfWfyxpI3k7mzQjYO81M/VZhf5eQ1mcyNKVmCaTJmUMlrML9ey/8n+IMj0YKDyKyyZTI/FKzTTDJD1Erey+aHrHQJSBvNDzJryMoy877y95fHotkt5oA1qwvZAOSNJW8m880nkm2jnoy0X9pu7feT26T6m6yfLVlIUEn7rVWTbPDWsEqGr6zWbwTyDShSEM10FSusQWmSNpuFg/mBa3ZhBIesvJ6Z/mCg5BRTwEqlJG8q2VitgSRk4za/kso6qSTU17ng15dq5DXJazCnUK9xpskHgfn+kPaEa5MsT9R7yAxZa9ukLWaFLe9xaxEhy8yQjfSBQbNLTL/kMvsl5c0WauOQN5m8CeWNlqxvNrOSC1Vpx7pMBbNqmoz8veXvLu2S68EfDGY4SPsLCgqMW9WZrN3m+yZcuArz/vJaElkRSpVtfiiEYr4GWS5tlb9xpNdF6SfcL7mmFbCRQkbeZFIBJOvXa2mfiDYoY71frCRk5GvyZCFrbtgSbEICybrhm7ebl6pNtd2Tkdcgryv4AyMR5O8ok4StnDlJ2mQNUvNDjiE7+8Q1YGWjkQ01VKUkzD3y8iabiWopFubGEAvZcGYqqKJlhqlcWgM2mYIqnVlDVq7T7BDXcxHIxiqThGhwSMm8VC6yPFnDVcgOCjOAomFWJ8lKXpNs2GYFJUd2SDeNhC7DVT35O8vffaaP4aXkFPPZtCRIZUM2N2IzrKTrQEj3gGzY5u1EROkqrl0EJglZqWKt/WxmhWd+hebXUiJKd0oC1krC1lqtWitchiwRpbO49sGGEtwVIPPmzpbgKpeIaDaIW8CGEhyycklENFvErYsgEglWmdhNQETpSHkXQSRSyTJciWi2mZGAJSKajRiwRESKMGCJiBSZkZ1cKkinMhHRdIXaORUt5T80ICKarRJ6FAER0WzEgCUiUoQBS0SkCAOWiEgRBiwRkSIMWCIiRRiwRESKMGCJiBRhwBIRKcKAJSJShAFLRKQIA5aISBGe7MVCBmY8ceLEZWOHyYgMpaWlxhwR0UQ8m9YUyPDi4UjAMmSJKBQG7CQOHDigj3wbC6lwly1bhqqqKuMWIppNeLrCGEl4Tka6FA4ePKh3MRARmRiwEchIuFu2bJly10Bw3y0RzW4M2DDMcJUKduXKlex/JaKoMWBDMMPVxJAlolgwYINImMrRBKG+7suOLAlaIqKpYMAGkWCV6nX79u3GLQEyL5PszIqNH94D/RjyGbPnfRjaNwjvOWOeiNIOA3bGeNHXsAGvvGsk7J+G0PXYBvSeDMwSUfphwM4YB6r3HkdrcWZgNmM1mj4cxsbbArNElH4YsBFIV4E58RAsIooWA9YgRw5YSaBap6kIfgwimt1m5KeyElBT+UVUosnPZeVkL7GQowsYsESzU1zPRSA/Ce3t7dWvW48XDcfcI79ixQoUFBTo14mI0kVcz0UggSlV6VTCVZjrxXoyFSKiVBR1wJr9kXLQfTBZZj3hiVw311+4cKF+ac4TEaW7uOzkMqtZqWzNKtU8ML+np0efT4U+WCKieIpLwEp4SpjKjh4JW7kuFatc8hypRDRbxRywoSpSa7eBBKt1HbOLgIhotogqYKU6lS4ACc5QJz2J1A0ghzDJcukyYD8sEc0GUQdspErU7H8NF6By+5kzZ4w5IqL0FlXAShVq9rMGD49izss5U83lu3btGg9b8yxUJSUlEStdIqJ0EXUfrBmOZiUqISpHEEj1KpdmP6y50ys4TBmuU+frdiGraSgw80kXXFmNMOaIKAXEvJPLJBWrBKn5YwKpcl0ulz6l2k9HpeqWPmL50LBO5qFmRETRmHbASkUqIWutTOXnsKn4k1jzuF3pzrBOchtDloiiFVMXgUzRnhTF7IONdxfB6OiocW165EQvkUjAmpV58CTVu1S6kQ2hMcuFtj1tcK3IRlZWNvIe64DHOqLB6T7Ur80JLFu3E/u9fmPBJZ636lF2e5a2Tg7Knu6D97yxAH64X96AImNZ0SNtGBozFhFRQsRUwVp/UDAVsp70zUoYxdvu3bvx29/+1phLDPlbyAdI8I6/y/Wja8iBhr3HMXK4GzV/akTxvwxq0ag578bOh2rhLenE8MnjeOfbTrjfGtTvdUkb2t67Cy8OjOD4oVbkvrsBm97w6kvG+rTg/ZEdDT8bwcjJd9BwUxfK6nrBjCVKnJgCVroEJDSnGphS4ckPD1R1G7S0tCQ8ZIV5xER4+aiuKoT9Gu3qfCcq/7EGi157G4OSsMM92OGvweYKJzLmArYbC7V18/V7XVKNhqe0+9u05Zn52PxUJYZe6oVbW+L/4yj8t61A7nxtZm4G8td/H53336rfi4gSY1p9sPH+uh8rCddkCdnItGTUwnNcdi5Ww4vRs4Dv1x4tIB2wW5ZfPVdbP4jNstzmcCL/9B/0Cjiz4D6UH9uAvLW1aOzUQvecHfkFDmQEViWiBJj2Tq5kkToha3HBH+geiIfFq9H05jD2PHMvHKM/R31pDspedsfv8YkoamkTsCnptBdDsGOR9rU+8yYHbEc8lp1WwKfnL49Hv2W53+PG4A3XaXWxH94jgxj6BLAvL0T5phZ0/6AGvqYevfuAiBIjbQL2c5/7HGpra/XL5NWPtpf6oR8c4Pegq6kRYw+XIl96AnJKUWPbgeZON8a0EPWf7MXOXZfv5Kp/NnB/v28Qzc92IHd9CZxaxI69W4+Khi64Za+WFsyeDw7Du2QRuwiIEigtAjY1wlXkY/UyN+rvykZWdjE6Mhuw69Fc6ZkF5jqx8dUW2HsrkLM0G3c+70XuNy7fyVX9lf3YtCIL2XkbMHR3O1rvs+tLnOs60f6lAWy6KwtZS3NQ0e9A+6vVcOhLiSgRZmTQQ5VaW1uxZs2aaYerHAc73SFt5KiK8EdKyHGwO+AYaEf5YuMmIkoLcR2TK5msXbs2BSpXIpqNUj5gFy1aZFybnnicNyHVzr1ARGqlfBdBPEk3QbQ/ATbJr9sYsESzU7guAgYsEdE0pW0fLBFRsmLAEhEpwoAlIlKEAUtEpAgDlohIEQYsEZEiDFgiIkUYsEREioT9ocGn/9mBP/zvRwLXHzumX8bLnDlzcO2112LevHmYO9d6in+aKDCQoevf+uG7owmH2suRaSwhouQR9S+5fvv4XwEX/qxfj3fAmiRgFyxYYMwlngxaKD+VDR5bS4bGkXHIZtyxnfha+WFs7GlF+Y1Bw80QUdKIPmD/SUbmC1AVsFLJ2u2B85kmg0iDOErAznjIvt+IrO87WLkSJbmk/KnsX/7yF+Na7EZHR41r0yMneomkp6dHD+BQk4yau2vXLmPNcOR8sC60de9E2e1ZcHX79Fv9RztQuzYHWVlZyFlbj97f6DfD1+1CVnkb0L8Vedoyc32cc6NjcxlytNuybi9D/Z7AsN1iqElbr70LO+XxXF3Q7zHZ+p2D6H06sDx7hda+o5Zhas550GUsy/qbPLi+NwifZcga775GuFZka23PRt66NgxNZYzw032oL89Dtv58E9sjIj2m9616/W8XaEsXmiuy0Pi+sdDoTimS5Vk5KHpkiu0hUijld3Lt3r074QMdSpfCwYMH9S6GyPrR3AfU/XgYLcVaTXq6Cxv+x36s2H4IIyePY8/DQHNlI4a0jMssa8dIVzVQ2IRDIyNoL9PWP+9F16Nl2P8V7baTIzjesw54/gEtZC6FYv8LbwNP7sFwSykyp7L+tlfwh/J2DMvzbwQaN7YZ43j50FtbjLZrNuMd7b4jA+246wMXNrwRCET/YCMeaALW/eS41vZDaM8fQFmtEephedD2UC28azq15xvBcEc5vE8+gJ1HA0sjPab/SDOqtnhR2jGMkQ8P4blb3eizjKgz1qeF74/saPiZ1taT76Dhpi6U1fWCGUuJlBZHESTLaLLBfbeXc6Ju60bk3piBDJtWQf6wHr7H61C+RPpXbbCvrkT1/A7sD9cjM9yB+jN1qLvfoQ/fbbthNSqrMtDxi0tDGzq1x9v4ZTsy5AmmuH7l8gz9+R1334PC04dxShLt1Nt4ZV856jblI0P6fjOcqNzagnwt7sa0qeeVNhQ+tRn587VlczPgrHgQlf1vYzBiwo5h9JQDK3KN9iwpRUNHk/YYEviRH9Pd1wb/+s2ovDUwyljm3TWoLtSv6vx/HIX/thXINe6bv/776Lz/1sBCogRJi4BNnSG7M2Eb79r2wXNS+1rbUKR3DwSmItRrX9H9F4xVgvh+7YH/aD2KxtfPQlGDG37Lt/rMa64zrkW/vulTef6zXrjznZDsH7dkNTZX5CIDXnj6gQ6XfJU3HjvbhQ5tFf2+YTlR+u0M7FijfcV/ug1dB7y4bnk+cm+QJ4n0mIG/1ZLPhz+5embBfSg/tgF5a2vR2NkL9zk78gscHPSREiptjoNNnZCdqGSH9pV3RPtaa5nqvmwsDKW4FcNB649szTUWhhDt+lNmx+Z/D3rckcnGG7PB6erE8C86UXOHDe7XKpD3t1vR94mxOKbHNCxejaY3h7HnmXvhGP056ktzUPay9mFiLCZKhLQJ2NSTAfuNNvS9FxQClp1IwTIW22HbNwC39Q5xXH+C+XY4B904Zb2vz43BI16tvZmw3+bFfu36BJM99lkPBg94MDbfgdziSjS8shfPLevCGwekXyHSY2bCsdSGUx+H26Hph/fIIIa0oLYvL0T5phZ0/6AGvqYeoz+ZKDHSJmBTZ+hukw35f18D+2v1aNwX6Lj0n+5H/b216AvTj2n7ajlqPt+B+qb+wN58vxf9zxaj9q3Qd4h2/QmW3IN1d3eh8cVBjMl9x9zoqCtD8zFpuR2lVSVwb6tHx7HAbqSxox1w3bdzYphf5hTeeORbaDZeL84O4cMPbFg0T7oIIj9mbkkNbC81jy/z7duBtn79qsaGsXfrUdHQBbcsPu+H54PD8C5ZxC4CSqi0CNjUC1fDrRvRvbsco02r9D7HnPLXseipp7A63EGvc53YuLsb5R83YtXSLP2wq9cX1OGpojB3iHb9CTJR0rIH1eeacafc9656HC7sRPt9geOWM77Rgp9us6OnOk9v+53PeHDPtmo4rX22weavxnM/rgZe+qZ+mFbW13dgdH03GlYHYjDiY2p/q13btWWVOfphWk8cc2J1vn43nXOd1rYvDWDTXdrjLs1BRb8D7a9Ww2EsJ0qEhP7QQNx4443Gtdi0trZizZo10w5XOQ62vb3dmIuNHBNbUFBgzM1Sn3TBtWIrxovLcYVoGphif+qU+NDlyoPnW5P0WRPNgLQdk2vt2rWpV7mms8XlaJ+wkyrKnVVEaSSmgL1i7L9w5fE3L5sSYdGi8IfuRCMeQ25z2G4isoqpi+Dqfw19APeF7Hvx56JGY25qpttFEE/STSAne4nFypUrGbBEs1TcTvYilepVP60z5i53MeML2vTXxtxEF75wBy58daMxF5BMAUtEFIu498EGgvTySVwx9t8hp6sO7cSc0/+pr0NElO5irmD9D+0dD9SpkH5b26ur9C4E6UowsYIlolSXtkcREBElKwYsEZEiDFgiIkUYsEREijBgiYgUYcASESkSc8DO+a/39GNapzxp6xMRzSZRHwdrHs8aq+DjZ5PpOFgZtFB+Khs8ttb1118flyG7ZRTXMnTHaUSB6CX6+YnSVdx+KiskZGOpSK0/MDAlU8DK6QbDkYCdbsgyYInSU1wDNp6mG7Cjo6NxOaPWdM4HKxXusmXLUFVVZdxyucasLLQZ14FqdI/UIRd+eLobsfWFLn24k8wvl+N/vdCA1TcE1pJA3LGoCbf/pBHN19XhUHs5Mk/3of6xJ9BxxK+PrLr5S4ex9cg9gWVyp3NudHynHs3/Zwhj83NR+cyLaCj2ac9fFuL5I5Dn+afvoOt9H7A4F+VPyuMETrYtvPsaUf90B/ql3XdvRmtzNXKN4QO8b9Vjk7ZsyJ+JwvWb4Xx3K/ybzPO2+uF+uRabXuqD52wGHEU1aNp+6b5EqShtf8m1e/fuhA90KF0KBw8e1LsYwqkbGUH3w9qVh7UK0gg3Gcu/+CUbNv/kuHbbcXSv8aL2sQ5YR6Xqf+Ft4Mk9GG4pReZ5N3Y+tAHuwnYcOj6Md77thLvXcmrr8150PVqG/V9pwqGTIzjesw54/gE0vu8M+fzhedD2UC28azoxrD3OcEc5vE8+gJ1HA0v9g414oAlYJ+0+eQjt+QMoq+2CDATjP9KMqi1elHYMY+TDQ3juVjf6BgP3E/Kay35kR8PPRrT7voOGm7pQVteLwEAwROklLY4iSJbRZIP7bieTUaBVpW/WIV/G8tfYC+9B/pH9OGIZMsv5eB02ftmOjAwbMNyDHR9Xo25DLjJtNthuLETN+kJjTc1wB+rP1KHufgdscwHbDatRWZWBjl9EO/TfGEZPObAi13icJaVo6GjS2ikDbvnQ80obCp/aHGj33Aw4Kx5EZf/bGNTa7e5rg1+rWitvDZSkUmVXW5ro/+Mo/LetQK5x3/z130fn/aFPf0mU6tIiYFN1yG5omen9v1pFtyJbH4Mqyxhq5dMLgcUi85rrjGtatP3aA/8dDti10AtFX360HkXyWMZU1OCGP+JAhKE4UfrtDOxYkwfX023oOuDFdcvzkXuDDI7lhUdrZIfLaLNM2S50aEs+veCD56QfSz4fvssms+A+lB/bgLy1tWjs7IX7nB35BQ4OTkhpKW2Og03FkPW9WRvoIujRvk7LsCoDTbAUe7EpbsXw+DAtxhT1Ti0bnK5ODP+iEzV32OB+rQJ5f7sVfZ8Yi2HH5n8Peo6pDgmzeDWa3hzGnmfuhWP056gvzUHZy0FDlxOlibQJ2FTk1UrBwvXrkJ9pDMV6XnYBhZe5xAmbx4dRGUY7hIzFdtj2DUwcOjvMuhGd9WDwgAdj8x3ILa5Ewyt78dyyLrxxQPouMmG/zYv9R6w9xRr9eTLhWGrDqY9H9Zsu54f3yKC+Q8++vBDlm1rQ/YMa+Jp6EG0nBlEqSJuATYWhuzNvygdOn4ZP+87u1wIpY4ED/W9p4eLT5n1udDQ1wrI/6HLL70GNbQeaO92Bx/hNP3a8dGknl+2r5aj5fAfqm/rhk8Dze9H/bDFq3wp06gY/f3in8MYj30LzPqMz+OwQPvzAhkXz5IPAjtKqEri31aPjWGDX1NjRDrju26kHe25JDWwvNY8v8+3bgbbxJtow9m49Khq64JbF5/3wfHAY3iWL2EVAaSktAjYVwlXYCx9E+a+fQF72BvSdBRyuXWj9wtuoyMtGTmkzxv5uM8rhg/+ccYdgc53Y+GoL7L3aV/bsHNz5z244v6GFpkmW7+5G+ceNWLU0C1m3l+H1BXV4qkg/gOuy5w9r/mo89+Nq4KVvIlv6WL++A6Pru9GwOhCDGd9owU+32dFTnaf3wd75jAf3bKuGU/L31o3YtV1bVpmDrL/JwxPHnFhtaaJzXSfavzSATXdpj7s0BxX9DrS/Wg2HsZwonaT8cbCtra1Ys2bNtMN1OsfBmuSHCgUFBcacIlJ5WnZyeV4rRtHHDdH3s37SBZexU22iQjQNxHOIbR+6XHnwfMs8DpYo/aTtcbBr165N+so1bvxD2HmvCx1HA1+//ae60Pj8GKq/7tTno7K4HO0TdlJFubOKiCaV0ICdM2f6Tx+PX3GJeAy5rXzYblsuql+4B57vrtK/uudUdsG+7YdaZWjsJCOipBK+i+DxvwIu/Fm/rqqLYN68eViwYIExl3jSTSAne4nFypUr1QcsESWlqM9F8P8OvY6xNzbo1+MdsFK5XnvttXrAzp0b5qh5IqIUEXXAEhHR1KTtTi4iomTFgCUiUoQBS0SkCAOWiEgRBiwRkSIMWCIiRRiwRESKMGCJiBRhwBIRKcKAJSJShAFLRKQIA5aISBEGLBGRIgxYIiJFGLBERIowYImIFGHAEhEpwoAlIlKEAUtEpAgDlohIEQYsEZEiDFgiIkUYsEREijBgiYgUYcASESnCgCUiUoQBS0SkCAOWiEgRBiwRkSIMWCIiRRiwRESKMGCJiBRhwBIRKcKAJSJShAFLRKQIA5aISBEGLBGRIgxYIiJFGLBERIowYImIFGHAEhEpwoAlIlKEAUtEpAgDlohIEQYsEZEiDFgiIkUYsEREijBgiYgUYcASESnCgCUiUoQBS0SkCAOWiEgRBiwRkSIMWCIiRRiwRESKMGCJiBRhwBIRKcKAJSJShAFLRKQIA5aISBEGLBGRIgxYIiJFGLBERIowYImIFGHAEhEpwoAlIlKEAUtEpAgDlohIEQYsEZEiDFgiIkUYsEREijBgiYgUYcASESnCgCUiUoQBS0SkCAOWiEgRBiwRkSIMWCIiRRiwRESKMGCJiBRhwBIRKcKAJSJShAFLRKQIA5aISBEGLBGRIgxYIiJFGLBERIowYImIFGHAEhEpwoAlIlIC+P9Hkl0aIyXdRwAAAABJRU5ErkJggg==)\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "LlwTAVzWA-zp"
      },
      "source": [
        "## Calculate Metrics"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "6d-D639O-iyM"
      },
      "outputs": [],
      "source": [
        "#@title ## Structure Metrics\n",
        "\n",
        "#@markdown Requires PDB uploaded files. All metrics will execute on all uploaded PDB files.\n",
        "\n",
        "#@markdown __MIF-ST__\n",
        "\n",
        "\n",
        "#@markdown __AlphaFold2 pLDDT__\n",
        "\n",
        "#@markdown Average taken from the b-factors in AlphaFold2-produced pdb files.\n",
        "\n",
        "#@markdown \n",
        "\n",
        "#@markdown ### __Metrics__\n",
        "# CHECKBOXES\n",
        "\n",
        "MIF_ST = True #@param {type:\"boolean\"}\n",
        "AlphaFold2_pLDDT = True #@param {type:\"boolean\"}\n",
        "\n",
        " \n",
        " # MIF-ST\n",
        "\n",
        "if MIF_ST: \n",
        "  with tempfile.TemporaryDirectory() as output_dir:\n",
        "    spec_file_path = output_dir + \"/spec_file.tsv\"\n",
        "    with open(spec_file_path, 'w') as f:\n",
        "      f.write('name\\tsequence\\tpdb\\n')\n",
        "      for pdb_file in glob(\"/content/pdbs/*.pdb\"):\n",
        "        seq = get_pdb_sequence(pdb_file) \n",
        "        name = Path(pdb_file).stem\n",
        "        f.write(name + '\\t' + seq + '\\t' + pdb_file + '\\n')\n",
        "    #print(spec_file_path)\n",
        "    proc = subprocess.run(['python', \"/tmp/extract_mif.py\", \"mifst\", spec_file_path, output_dir + \"/\", \"logits\", \"--include\", \"logp\", \"--device\", device], stdout=subprocess.PIPE, stderr=subprocess.PIPE, check=True)\n",
        "    # print(proc.stderr.decode(\"utf-8\"))\n",
        "    # print(proc.stdout.decode(\"utf-8\"))\n",
        "    df = pd.read_table(output_dir + '/mifst_logp.tsv')\n",
        "    df = df.rename(columns={'name': 'id', 'logp': 'mifst_logp'}, )\n",
        "    for _, row in df.iterrows():\n",
        "      add_metric(results, row[\"id\"], \"MIF-ST\", row[\"mifst_logp\"])\n",
        "\n",
        "\n",
        "  \n",
        "# pLDDT\n",
        "if AlphaFold2_pLDDT:\n",
        "  for pdb_file in glob(\"/content/pdbs/*.pdb\"):\n",
        "    fstem = Path(pdb_file).stem\n",
        "    name = fstem\n",
        "    pdb_file = pdb.PDBFile.read(pdb_file)\n",
        "    atoms = pdb_file.get_structure(extra_fields = ['b_factor'])\n",
        "    prev_residue = -1\n",
        "    plddt_sum = 0\n",
        "    residue_count = 0\n",
        "    for a in atoms[0]:\n",
        "      if a.res_id != prev_residue:\n",
        "        prev_residue = a.res_id\n",
        "        residue_count += 1\n",
        "        plddt_sum += a.b_factor\n",
        "    add_metric(results, name, \"AlphaFold2 pLDDT\", plddt_sum/residue_count)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "IwJ-zDe7HFad"
      },
      "outputs": [],
      "source": [
        "#@title Single-sequence metrics\n",
        "\n",
        "#@markdown __CARP-640m__\n",
        "\n",
        "\n",
        "#@markdown ### __Metrics__\n",
        "# CHECKBOXES\n",
        "CARP_640m_logp = True #@param {type:\"boolean\"}\n",
        "\n",
        "\n",
        "target_seqs_file = \"/tmp/target_seqs.fasta\"\n",
        "with open(target_seqs_file,\"w\") as fh:\n",
        "  for target_fasta in glob(\"/content/target_seqs/*\"):\n",
        "    for name, seq in zip(*parse_fasta(target_fasta, return_names=True, clean=\"unalign\")):\n",
        "      print(f\">{name}\\n{seq}\", file=fh)\n",
        "\n",
        "#CARP\n",
        "if CARP_640m_logp: \n",
        "  with tempfile.TemporaryDirectory() as output_dir:\n",
        "    proc = subprocess.run(['python', \"/tmp/extract.py\", \"carp_640M\", \"/tmp/target_seqs.fasta\", output_dir + \"/\", \"--repr_layers\", \"logits\", \"--include\", \"logp\", \"--device\", device], stdout=subprocess.PIPE, stderr=subprocess.PIPE, check=True)\n",
        "    # print(proc.stderr)\n",
        "    # print(proc.stdout)\n",
        "    df = pd.read_table(output_dir + '/carp_640M_logp.tsv')\n",
        "    df = df.rename(columns={'name': 'id', 'logp': 'carp640m_logp'},)\n",
        "    for _, row in df.iterrows():\n",
        "      add_metric(results, row[\"id\"], \"CARP-640m\", row[\"carp640m_logp\"])\n",
        "\n",
        "\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 17
        },
        "id": "ZNLh4J1H8h4J",
        "outputId": "f7246485-e283-4485-8343-252553e5fb21"
      },
      "outputs": [
        {
          "data": {
            "application/javascript": "\n    async function download(id, filename, size) {\n      if (!google.colab.kernel.accessAllowed) {\n        return;\n      }\n      const div = document.createElement('div');\n      const label = document.createElement('label');\n      label.textContent = `Downloading \"${filename}\": `;\n      div.appendChild(label);\n      const progress = document.createElement('progress');\n      progress.max = size;\n      div.appendChild(progress);\n      document.body.appendChild(div);\n\n      const buffers = [];\n      let downloaded = 0;\n\n      const channel = await google.colab.kernel.comms.open(id);\n      // Send a message to notify the kernel that we're ready.\n      channel.send({})\n\n      for await (const message of channel.messages) {\n        // Send a message to notify the kernel that we're ready.\n        channel.send({})\n        if (message.buffers) {\n          for (const buffer of message.buffers) {\n            buffers.push(buffer);\n            downloaded += buffer.byteLength;\n            progress.value = downloaded;\n          }\n        }\n      }\n      const blob = new Blob(buffers, {type: 'application/binary'});\n      const a = document.createElement('a');\n      a.href = window.URL.createObjectURL(blob);\n      a.download = filename;\n      div.appendChild(a);\n      a.click();\n      div.remove();\n    }\n  ",
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        },
        {
          "data": {
            "application/javascript": "download(\"download_e6c32fec-10c9-4496-8f91-332029225fc6\", \"calculated_metrics.csv\", 735)",
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        }
      ],
      "source": [
        "#@title Download results\n",
        "# OPTION FOR \n",
        "df = pd.DataFrame.from_dict(results, orient=\"index\")\n",
        "df.to_csv(f\"/tmp/calculated_metrics.csv\")\n",
        "\n",
        "files.download(f\"/tmp/calculated_metrics.csv\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "sZ5vnK82W9I9"
      },
      "source": [
        "### Missing metrics\n",
        "Phobius and Rosetta are discussed in the paper, but not available here because of licensing issues."
      ]
    }
  ],
  "metadata": {
    "accelerator": "GPU",
    "colab": {
      "authorship_tag": "ABX9TyM1pkRWND79sLDeeH9a5W4Y",
      "include_colab_link": true,
      "provenance": [],
      "toc_visible": true
    },
    "gpuClass": "standard",
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}
